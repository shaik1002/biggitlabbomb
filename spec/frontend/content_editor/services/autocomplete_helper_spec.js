import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import AutocompleteHelper, {
  defaultSorter,
  customSorter,
  createDataSource,
} from '~/content_editor/services/autocomplete_helper';
import {
  MOCK_MEMBERS,
  MOCK_COMMANDS,
  MOCK_EPICS,
  MOCK_ISSUES,
  MOCK_LABELS,
  MOCK_MILESTONES,
  MOCK_SNIPPETS,
  MOCK_VULNERABILITIES,
  MOCK_MERGE_REQUESTS,
  MOCK_ASSIGNEES,
  MOCK_REVIEWERS,
  MOCK_WIKIS,
} from './autocomplete_mock_data';

jest.mock('~/emoji', () => ({
  initEmojiMap: () => jest.fn(),
  getAllEmoji: () => [{ name: 'thumbsup' }],
}));

describe('defaultSorter', () => {
  it('returns items as is if query is empty', () => {
    const items = [{ name: 'abc' }, { name: 'bcd' }, { name: 'cde' }];
    const sorter = defaultSorter(['name']);
    expect(sorter(items, '')).toEqual(items);
  });

  it('sorts items based on query match', () => {
    const items = [{ name: 'abc' }, { name: 'bcd' }, { name: 'cde' }];
    const sorter = defaultSorter(['name']);
    expect(sorter(items, 'b')).toEqual([{ name: 'bcd' }, { name: 'abc' }, { name: 'cde' }]);
  });

  it('sorts items based on query match in multiple fields', () => {
    const items = [
      { name: 'wabc', description: 'xyz' },
      { name: 'bcd', description: 'wxy' },
      { name: 'cde', description: 'vwx' },
    ];
    const sorter = defaultSorter(['name', 'description']);
    expect(sorter(items, 'w')).toEqual([
      { name: 'wabc', description: 'xyz' },
      { name: 'bcd', description: 'wxy' },
      { name: 'cde', description: 'vwx' },
    ]);
  });
});

describe('customSorter', () => {
  it('sorts items based on custom sorter function', () => {
    const items = [3, 1, 2];
    const sorter = customSorter((a, b) => a - b);
    expect(sorter(items)).toEqual([1, 2, 3]);
  });
});

describe('createDataSource', () => {
  let mock;

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    mock.restore();
  });

  it('fetches data from source and filters based on query', async () => {
    const data = [
      { name: 'abc', description: 'xyz' },
      { name: 'bcd', description: 'wxy' },
      { name: 'cde', description: 'vwx' },
    ];
    mock.onGet('/source').reply(200, data);

    const dataSource = createDataSource({
      source: '/source',
      searchFields: ['name', 'description'],
    });

    const results = await dataSource.search('b');
    expect(results).toEqual([
      { name: 'bcd', description: 'wxy' },
      { name: 'abc', description: 'xyz' },
    ]);
  });

  it('handles source fetch errors', async () => {
    mock.onGet('/source').reply(500);

    const dataSource = createDataSource({
      source: '/source',
      searchFields: ['name', 'description'],
      sorter: (items) => items,
    });

    const results = await dataSource.search('b');
    expect(results).toEqual([]);
  });
});

describe('AutocompleteHelper', () => {
  let mock;
  let autocompleteHelper;
  let dateNowOld;

  beforeEach(() => {
    mock = new MockAdapter(axios);
    const dataSourceUrls = {
      members: '/members',
      issues: '/issues',
      snippets: '/snippets',
      labels: '/labels',
      epics: '/epics',
      milestones: '/milestones',
      mergeRequests: '/mergeRequests',
      vulnerabilities: '/vulnerabilities',
      commands: '/commands',
      wikis: '/wikis',
    };

    mock.onGet('/members').reply(200, MOCK_MEMBERS);
    mock.onGet('/issues').reply(200, MOCK_ISSUES);
    mock.onGet('/snippets').reply(200, MOCK_SNIPPETS);
    mock.onGet('/labels').reply(200, MOCK_LABELS);
    mock.onGet('/epics').reply(200, MOCK_EPICS);
    mock.onGet('/milestones').reply(200, MOCK_MILESTONES);
    mock.onGet('/mergeRequests').reply(200, MOCK_MERGE_REQUESTS);
    mock.onGet('/vulnerabilities').reply(200, MOCK_VULNERABILITIES);
    mock.onGet('/commands').reply(200, MOCK_COMMANDS);
    mock.onGet('/wikis').reply(200, MOCK_WIKIS);

    const sidebarMediator = {
      store: {
        assignees: MOCK_ASSIGNEES,
        reviewers: MOCK_REVIEWERS,
      },
    };

    autocompleteHelper = new AutocompleteHelper({
      dataSourceUrls,
      sidebarMediator,
    });

    dateNowOld = Date.now();

    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2023-11-14').getTime());
  });

  afterEach(() => {
    mock.restore();

    delete gl.GfmAutoComplete;

    jest.spyOn(Date, 'now').mockImplementation(() => dateNowOld);
  });

  it.each`
    referenceType      | query
    ${'user'}          | ${'r'}
    ${'issue'}         | ${'q'}
    ${'snippet'}       | ${'s'}
    ${'label'}         | ${'c'}
    ${'epic'}          | ${'n'}
    ${'milestone'}     | ${'16'}
    ${'merge_request'} | ${'n'}
    ${'vulnerability'} | ${'cross'}
    ${'command'}       | ${'re'}
    ${'wiki'}          | ${'ho'}
  `(
    'for reference type "$referenceType", searches for "$query" correctly',
    async ({ referenceType, query }) => {
      const dataSource = autocompleteHelper.getDataSource(referenceType);
      const results = await dataSource.search(query);

      expect(
        results.map(({ title, name, username }) => username || name || title),
      ).toMatchSnapshot();
    },
  );

  it.each`
    referenceType | command
    ${'label'}    | ${'/label'}
    ${'label'}    | ${'/unlabel'}
    ${'label'}    | ${'/relabel'}
    ${'user'}     | ${'/assign'}
    ${'user'}     | ${'/reassign'}
    ${'user'}     | ${'/unassign'}
    ${'user'}     | ${'/assign_reviewer'}
    ${'user'}     | ${'/unassign_reviewer'}
    ${'user'}     | ${'/reassign_reviewer'}
  `(
    'filters items based on command "$command" for reference type "$referenceType" and command',
    async ({ referenceType, command }) => {
      const dataSource = autocompleteHelper.getDataSource(referenceType, { command });
      const results = await dataSource.search();

      expect(
        results.map(({ username, name, title }) => username || name || title),
      ).toMatchSnapshot();
    },
  );

  it('loads default datasources if not passed', () => {
    gl.GfmAutoComplete = {
      dataSources: {
        members: '/gitlab-org/gitlab-test/-/autocomplete_sources/members',
      },
    };
    autocompleteHelper = new AutocompleteHelper({});

    expect(autocompleteHelper.dataSourceUrls.members).toBe(
      '/gitlab-org/gitlab-test/-/autocomplete_sources/members',
    );
  });

  it("loads emoji if dataSources doesn't exist", async () => {
    autocompleteHelper = new AutocompleteHelper({});

    const dataSource = autocompleteHelper.getDataSource('emoji');
    const results = await dataSource.search('');

    expect(results).toEqual([{ emoji: { name: 'thumbsup' }, fieldValue: 'thumbsup' }]);
  });
});
