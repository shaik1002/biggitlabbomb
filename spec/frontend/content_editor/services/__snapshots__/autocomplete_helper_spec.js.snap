// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteHelper filters items based on command "/assign" for reference type "user" and command 1`] = `
Array [
  "florida.schoen",
  "root",
  "all",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
  "jashkenas",
  "twitter",
]
`;

exports[`AutocompleteHelper filters items based on command "/assign_reviewer" for reference type "user" and command 1`] = `
Array [
  "florida.schoen",
  "root",
  "all",
  "errol",
  "evelynn_olson",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
  "jashkenas",
  "twitter",
]
`;

exports[`AutocompleteHelper filters items based on command "/label" for reference type "label" and command 1`] = `
Array [
  "Bronce",
  "Contour",
  "Corolla",
  "Cygsync",
  "Frontier",
  "Grand Am",
  "Onesync",
  "Phone",
  "Pynefunc",
  "Trinix",
  "Trounswood",
  "group::knowledge",
  "scoped label",
  "type::one",
  "type::two",
]
`;

exports[`AutocompleteHelper filters items based on command "/reassign" for reference type "user" and command 1`] = `
Array [
  "florida.schoen",
  "root",
  "all",
  "errol",
  "evelynn_olson",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
]
`;

exports[`AutocompleteHelper filters items based on command "/reassign_reviewer" for reference type "user" and command 1`] = `
Array [
  "florida.schoen",
  "root",
  "all",
  "errol",
  "evelynn_olson",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
]
`;

exports[`AutocompleteHelper filters items based on command "/relabel" for reference type "label" and command 1`] = `
Array [
  "Amsche",
  "Brioffe",
  "Bronce",
  "Bryncefunc",
  "Contour",
  "Corolla",
  "Cygsync",
  "Frontier",
  "Ghost",
  "Grand Am",
  "Onesync",
  "Phone",
  "Pynefunc",
  "Trinix",
  "Trounswood",
]
`;

exports[`AutocompleteHelper filters items based on command "/unassign" for reference type "user" and command 1`] = `
Array [
  "errol",
  "evelynn_olson",
]
`;

exports[`AutocompleteHelper filters items based on command "/unassign_reviewer" for reference type "user" and command 1`] = `
Array [
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
]
`;

exports[`AutocompleteHelper filters items based on command "/unlabel" for reference type "label" and command 1`] = `
Array [
  "Amsche",
  "Brioffe",
  "Bryncefunc",
  "Ghost",
]
`;

exports[`AutocompleteHelper for reference type "command", searches for "re" correctly 1`] = `
Array [
  "relabel",
  "remove_milestone",
  "remove_estimate",
  "remove_time_spent",
  "relate",
  "remove_epic",
  "reassign",
  "create_merge_request",
]
`;

exports[`AutocompleteHelper for reference type "epic", searches for "n" correctly 1`] = `
Array [
  "Nobis quidem aspernatur reprehenderit sunt ut ipsum tempora sapiente sed iste.",
  "Minus eius ut omnis quos sunt dicta ex ipsum.",
  "Quae nostrum possimus rerum aliquam pariatur a eos aut id.",
  "Dicta incidunt vel dignissimos sint sit esse est quibusdam quidem consequatur.",
  "Doloremque a quisquam qui culpa numquam doloribus similique iure enim.",
]
`;

exports[`AutocompleteHelper for reference type "issue", searches for "q" correctly 1`] = `
Array [
  "Quasi id et et nihil sint autem.",
  "Eaque omnis eius quas necessitatibus hic ut et corrupti.",
  "Aut quisquam magnam eos distinctio incidunt perferendis fugit.",
  "Dolorem quisquam cupiditate consequatur perspiciatis sequi eligendi ullam.",
  "Nesciunt quia molestiae in aliquam amet et dolorem.",
  "Porro tempore qui qui culpa saepe et nam quos.",
  "Sed sint a est consequatur quae quasi autem debitis alias.",
  "Molestiae minima maxime optio nihil quam eveniet dolor.",
  "Et laboriosam aut ratione voluptatem quasi recusandae.",
  "Et molestiae delectus voluptates velit vero illo aut rerum quo et.",
]
`;

exports[`AutocompleteHelper for reference type "label", searches for "c" correctly 1`] = `
Array [
  "Contour",
  "Corolla",
  "Cygsync",
  "scoped label",
  "Amsche",
  "Bronce",
  "Bryncefunc",
  "Onesync",
  "Pynefunc",
]
`;

exports[`AutocompleteHelper for reference type "merge_request", searches for "n" correctly 1`] = `
Array [
  "Blanditiis maxime voluptatem ut pariatur vel autem vero non quod libero.",
  "Optio nemo qui dolorem sit ipsum qui saepe.",
  "Draft: Alunny/publish lib",
  "Draft: Fix event current target",
  "Draft: Resolve \\"hgvbbvnnb\\"",
  "Autem eaque et sed provident enim corrupti molestiae.",
  "Always call registry's trigger method from withRegistration",
]
`;

exports[`AutocompleteHelper for reference type "milestone", searches for "16" correctly 1`] = `
Array [
  "16.7",
  "16.8",
  "16.9",
  "16.10",
  "16.11",
  "16.0",
  "16.1",
  "16.2",
  "16.3",
  "16.4",
  "16.5",
  "16.6",
]
`;

exports[`AutocompleteHelper for reference type "snippet", searches for "s" correctly 1`] = `
Array [
  "ss",
  "test snippet",
  "another test snippet",
]
`;

exports[`AutocompleteHelper for reference type "user", searches for "r" correctly 1`] = `
Array [
  "root",
  "errol",
  "lakeesha.batz",
  "myrtis",
  "florida.schoen",
  "laurene_blick",
  "all",
  "twitter",
  "gitlab-org",
  "evelynn_olson",
]
`;

exports[`AutocompleteHelper for reference type "vulnerability", searches for "cross" correctly 1`] = `
Array [
  "Cross Site Scripting (Persistent)",
  "Cross Site Scripting (Persistent)",
  "Cross Site Scripting (Persistent)",
]
`;

exports[`AutocompleteHelper for reference type "wiki", searches for "ho" correctly 1`] = `
Array [
  "Home",
  "How to use GitLab",
]
`;
